### ISSUE TITLE
Show thread number in messages

### ISSUE THREAD

**Issue by averissimo** [zatonovo/futile.logger#39]
Use case:

When using foreach with parallel library (%dopar%) it would be nice to have the thread number ou process where it is executing.

**Comment by r2evans** [comment]
Can you do something like `flog.debug(msg = "foobar", PID = Sys.getpid(), ...)`? I don't think making it a mandatory field in the log is required, and it is relatively straight-forward to include.

**Comment by averissimo** [comment]
How do I create a layout to include that PID variable? 

Using flog.debug / flog.info with default layout does not show PID

    > flog.debug(msg = 'yada', PID = Sys.getpid())
    DEBUG [2016-07-29 15:48:36] yada

And when using tracearg as layout, if it has a msg string it does not show additional variables.

    > flog.layout(layout.tracearg, name='tracer')
    NULL
    > flog.debug('yada', PID = Sys.getpid(), name='tracer')
    DEBUG [2016-07-29 15:50:41] yada

**Comment by r2evans** [comment]
The only things that are currently included in a template are the tilde-tokens as listed in `layout.format`. It **should be trivial** to add `~p` (or something similar) to the list of tokens. Perhaps this:

```diff
diff --git a/R/layout.R b/R/layout.R
index 2ff0c5f..44812ea 100644
--- a/R/layout.R
+++ b/R/layout.R
@@ -44,6 +44,7 @@
 #' \item{~n}{Namespace}
 #' \item{~f}{The calling function}
 #' \item{~m}{The message}
+#' \item{~p}{The process PID}
 #' }
 #'
 #' \code{layout.json} converts the message and any additional objects provided
@@ -141,6 +142,7 @@ layout.json <- function(level, msg, ...) {
 # ~n - Namespace
 # ~f - Calling function
 # ~m - Message
+# ~p - PID
 #
 # layout <- layout.format('[~l] [~t] [~n.~f] ~m')
 # flog.layout(layout)
@@ -155,6 +157,7 @@ layout.format <- function(format, datetime.fmt="%Y-%m-%d %H:%M:%S")
     the.namespace <- flog.namespace(.where)
     the.namespace <- ifelse(the.namespace == 'futile.logger', 'ROOT', the.namespace)
     the.function <- .get.parent.func.name(.where) 
+    the.pid <- Sys.getpid()
     #pattern <- c('~l','~t','~n','~f','~m')
     #replace <- c(the.level, the.time, the.namespace, the.function, msg)
     message <- gsub('~l',the.level, format, fixed=TRUE)
@@ -162,6 +165,7 @@ layout.format <- function(format, datetime.fmt="%Y-%m-%d %H:%M:%S")
     message <- gsub('~n',the.namespace, message, fixed=TRUE)
     message <- gsub('~f',the.function, message, fixed=TRUE)
     message <- gsub('~m',msg, message, fixed=TRUE)
+    message <- gsub('~p',the.pid, message, fixed=TRUE)
     sprintf("%s\n", message)
   }
 }
```

Though I don't use it personally, I believe that `layout.tracearg` will only tell you something about a single variable (name and contents), nothing else is used.

**Comment by averissimo** [comment]
Thanks a million!!

I have added this to my fork at branch pid :)

https://github.com/averissimo/futile.logger/tree/pid

**Comment by averissimo** [comment]
Yup! I besides @r2evans code for the token I also added a function to create a simple layout with pid.

also added documentation and tests

see #42

**Comment by iwccsbfb** [comment]
@averissimo @r2evans  Actually, system pid can be passed in logging in a simple way: 
` flog.info('%d msg', Sys.getpid())`

**Comment by averissimo** [comment]
Thanks, at the time I was actually using `paste0('.... ',var1,' ...')` and asked this without really looking at the documentation of futile.logger.. I'm really sorry about that :)

I kinda prefer the approach from @r2evans as it avoids having to create proxy functions for the `flog.info/debug/warn/error` -- it also avoiding a lot of repetitions in the code and across projects and is more robust over changes

I didn't see anywhere in the documentation but **it would be awesome to extend the layout.format to accept function arguments as tokens**, such as below, much like flog.**** already does :)

`layout.format('[~l] [~t] [~n.~f %f] ~m', function(args){Sys.getpid())`

*where args holds all information from tokens*
